### PR TITLE
Convert pal_gssapi to C

### PR DESCRIPTION
**Pull request by kasper3**
`ARRAY_SIZE` is defined for _cplusplus in pal_utilities.h based on https://blogs.msdn.microsoft.com/the1/2004/05/07/how-would-you-get-the-count-of-an-array-in-c-2/ (some article which talks about COUNTOF instead of ARRAY_SIZE which sounds like size in bytes..). for C, if we use something like `#define ARRAY_SIZE(a) sizeof(a)/sizeof(a[0])`, that only works for stack-allocated array, and not heap, re: https://stackoverflow.com/a/10349610.
I have used this approach assuming the allocation here is stack allocation.

### CODE REVIEW

**Comment by bartonjs** [comment]
We had reasons that we converted things originally in C to be C++.  @stephentoub I think you converted my original OpenSSL code.

My instinct is "this goes against the project standards".

**Comment by wfurt** [comment]
I think we are trying to get running on platforms without c++ @bartonjs 
cc: @janvorli

**Comment by kasper3** [comment]
@bartonjs, actually there is nothing that is requiring us to link every PAL so/dylib with libstdc++, if we test `ldd PAL*.so` and look at the code that rarely uses a `new` or `delete`. Therefore libstdc++ link is wasteful.
Secondly, this is not the first PR in the series of "convert to C". That work started to support Mono, that uses C to target wider variety of platforms (where C++11 etc. is not available). Another reason to consider.

> "this goes against the project standards"

except the stats suggest otherwise 😸 

image

**Comment by EgorBo** [comment]
We planned to reuse pal_gssapi in Mono too (e.g. to fix https://github.com/mono/mono/issues/9028) 
Great job @kasper3!

**Comment by janvorli** [comment]
@kasper3 there are some issues in the build:
```
16:00:06 /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/Native/Unix/System.Net.Security.Native/pal_gssapi.c:162:50: error: cast from 'const struct gss_OID_desc_struct *' to 'struct gss_OID_desc_struct *' drops const qualifier [-Werror,-Wcast-qual]
16:00:06         nameType = (struct gss_OID_desc_struct *)GSS_KRB5_NT_PRINCIPAL_NAME;
16:00:06                                                  ^
16:00:06 /mnt/j/workspace/dotnet_corefx/master/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/src/Native/Unix/System.Net.Security.Native/pal_gssapi.c:222:53: error: cast from 'const struct gss_OID_desc_struct *' to 'struct gss_OID_desc_struct *' drops const qualifier [-Werror,-Wcast-qual]
16:00:06     gss_OID krbMech = (struct gss_OID_desc_struct *)gss_mech_krb5;
16:00:06                                                     ^
16:00:06 2 errors generated.
```

**Comment by EgorBo** [comment]
I guess `(struct gss_OID_desc_struct *)` should be replaced with just `(gss_OID)`

**Comment by kasper3** [comment]
for some reason i can't reproduce this `const_cast` replacement error on alpine 3.7 locally using `build-native.sh`

File: src/Native/Unix/CMakeLists.txt [hunk 1]
```diff
@@ -21,6 +21,7 @@ add_compile_options(-Wno-unused-local-typedef)
 add_compile_options(-Wno-c99-extensions)
 add_compile_options(-Wno-c11-extensions)
 add_compile_options(-Wmissing-prototypes)
+add_compile_options(-Wno-incompatible-pointer-types-discards-qualifiers)
```
Comment (janvorli) [inline src/Native/Unix/CMakeLists.txt]: Is there no way to fix the code instead? The warning disabling should be the last resort.

Comment (kasper3) [reply src/Native/Unix/CMakeLists.txt]: In Alpine 3.7, Ubuntu 18 and many other os/distro/versions, that are shipping with clang 5+, this is not required. This is only for clang4 and below that assigning `const` to non-const raises this warning, and -werror renders it as error.

File: src/Native/Unix/System.Net.Security.Native/pal_gssapi.c [hunk 2]
```diff
@@ -21,61 +21,61 @@
...
+c_static_assert(PAL_GSS_C_DCE_STYLE == GSS_C_DCE_STYLE);
+c_static_assert(PAL_GSS_C_IDENTIFY_FLAG == GSS_C_IDENTIFY_FLAG);
+c_static_assert(PAL_GSS_C_EXTENDED_ERROR_FLAG == GSS_C_EXTENDED_ERROR_FLAG);
+c_static_assert(PAL_GSS_C_DELEG_POLICY_FLAG == GSS_C_DELEG_POLICY_FLAG);
+
+c_static_assert(PAL_GSS_COMPLETE == GSS_S_COMPLETE);
+c_static_assert(PAL_GSS_CONTINUE_NEEDED == GSS_S_CONTINUE_NEEDED);
 
 #if !HAVE_GSS_SPNEGO_MECHANISM
 static char gss_spnego_oid_value[] = "\x2b\x06\x01\x05\x05\x02"; // Binary representation of SPNEGO Oid (RFC 4178)
```
Comment (jaykrell) [inline src/Native/Unix/System.Net.Security.Native/pal_gssapi.c]: const?

File: src/Native/Unix/System.Net.Security.Native/pal_gssapi.c [hunk 3]
```diff
@@ -21,61 +21,61 @@
...
-                                              .elements = static_cast<void*>(gss_ntlm_oid_value)};
+                                              .elements = gss_ntlm_oid_value};
 #endif
 
 // transfers ownership of the underlying data from gssBuffer to PAL_GssBuffer
 static void NetSecurityNative_MoveBuffer(gss_buffer_t gssBuffer, struct PAL_GssBuffer* targetBuffer)
 {
-    assert(gssBuffer != nullptr);
-    assert(targetBuffer != nullptr);
+    assert(gssBuffer != NULL);
```
Comment (jaykrell) [inline src/Native/Unix/System.Net.Security.Native/pal_gssapi.c]: Could also just assert(gssBuffer).

File: src/Native/Unix/System.Net.Security.Native/pal_gssapi.c [hunk 4]
```diff
@@ -21,61 +21,61 @@
...
+    assert(targetBuffer != NULL);
 
-    targetBuffer->length = static_cast<uint64_t>(gssBuffer->length);
-    targetBuffer->data = static_cast<uint8_t*>(gssBuffer->value);
+    targetBuffer->length = (uint64_t)(gssBuffer->length);
```
Comment (jaykrell) [inline src/Native/Unix/System.Net.Security.Native/pal_gssapi.c]: No need for last set of parens on each line.

Comment (gfoidl) [reply src/Native/Unix/System.Net.Security.Native/pal_gssapi.c]: For readability I would keep them.

File: src/Native/Unix/System.Net.Security.Native/pal_gssapi.c (hunk 2 above)
Comment (kasper3) [reply src/Native/Unix/System.Net.Security.Native/pal_gssapi.c]: this triggers another instance of `incompatible-pointer-types-discards-qualifiers` at usage and I am not touching it anyway in PR.

File: src/Native/Unix/System.Net.Security.Native/pal_gssapi.c (hunk 3 above)
Comment (kasper3) [reply src/Native/Unix/System.Net.Security.Native/pal_gssapi.c]: we believe in consistency with all other places in this repo and this is the way it has been asserted everywhere..

File: src/Native/Unix/Common/pal_utilities.h [hunk 5]
```diff
@@ -148,6 +148,8 @@ inline void SafeStringCopy(char* destination, int32_t destinationSize, const cha
     }
 }
 
+#else // __cplusplus
+#define ARRAY_SIZE(a) (sizeof(a)/sizeof(a[0]))
```
Comment (stephentoub) [inline src/Native/Unix/Common/pal_utilities.h]: Is there any risk this gets used with an empty array?

Comment (kasper3) [reply src/Native/Unix/Common/pal_utilities.h]: I've just tested:

```sh
clang -x c -o test.exe - << EOF
#define ARRAY_SIZE(a) (sizeof(a)/sizeof(a[0]))
#include <stdio.h>
int main(void)
{
  static char first[] = {};
  static char second[] = "";
  static char third[] = "\0";
  printf("count of first %lu\n", ARRAY_SIZE(first));
  printf("count of second %lu\n", ARRAY_SIZE(second));
  printf("count of third %lu\n", ARRAY_SIZE(third));
  return 0;
}
EOF
```

produces:

count of first 0
count of second 1
count of third 2

Comment (stephentoub) [reply src/Native/Unix/Common/pal_utilities.h]: Ok. Thanks.

File: src/Native/Unix/System.Net.Security.Native/pal_gssapi.c [hunk 6]
```diff
@@ -251,82 +250,82 @@ extern "C" uint32_t NetSecurityNative_InitSecContext(uint32_t* minorStatus,
...
 
-extern "C" void NetSecurityNative_ReleaseGssBuffer(void* buffer, uint64_t length)
+void NetSecurityNative_ReleaseGssBuffer(void* buffer, uint64_t length)
 {
-    assert(buffer != nullptr);
+    assert(buffer != NULL);
 
     uint32_t minorStatus;
-    GssBuffer gssBuffer{.length = static_cast<size_t>(length), .value = buffer};
+    GssBuffer gssBuffer = {.length = (size_t)(length), .value = buffer};
```
Comment (stephentoub) [inline src/Native/Unix/System.Net.Security.Native/pal_gssapi.c]: Nit: `(length)` can be `length`

**Pull request merged**